src/core/lib/event_engine/windows/windows_endpoint.cc

```diff
@@ -68,76 +76,79 @@ WindowsEndpoint::~WindowsEndpoint() {
   }
   DWORD bytes_read = 0;
   DWORD flags = 0;
-  // First let's try a synchronous, non-blocking read.
+  // First try a synchronous, non-blocking read.
   int status =
       WSARecv(io_state_->socket->raw_socket(), wsa_buffers,
               (DWORD)buffer->Count(), &bytes_read, &flags, nullptr, nullptr);
   int wsa_error = status == 0 ? 0 : WSAGetLastError();
-  // Did we get data immediately ? Yay.
   if (wsa_error != WSAEWOULDBLOCK) {
+    // Data or some error was returned immediately.
     io_state_->socket->read_info()->SetResult(
         {/*wsa_error=*/wsa_error, /*bytes_read=*/bytes_read});
-    absl::Status result;
-    if (bytes_read == 0) {
-      result = absl::UnavailableError("End of TCP stream");
-      grpc_core::StatusSetInt(&result, grpc_core::StatusIntProperty::kRpcStatus,
-                              GRPC_STATUS_UNAVAILABLE);
-      buffer->Clear();
-    } else {
-      result = absl::OkStatus();
-      // prune slicebuffer
-      if (bytes_read != buffer->Length()) {
-        buffer->RemoveLastNBytes(buffer->Length() - bytes_read);
-      }
-    }
-    executor_->Run(
-        [result, on_read = std::move(on_read)]() mutable { on_read(result); });
-    return false;
+    executor_->Run(&io_state_->handle_read_event);
+    return absl::OkStatus();
+  }
+  // If the endpoint has already received some data, and the next call would
+  // block, return the data in case that is all the data the reader expects.
+  if (io_state_->handle_read_event.MaybeFinishIfDataHasAlreadyBeenRead(
+          absl::OkStatus())) {
+    return absl::OkStatus();
   }
   // Otherwise, let's retry, by queuing a read.
-  memset(io_state_->socket->read_info()->overlapped(), 0, sizeof(OVERLAPPED));
   status = WSARecv(io_state_->socket->raw_socket(), wsa_buffers,
                    (DWORD)buffer->Count(), &bytes_read, &flags,
                    io_state_->socket->read_info()->overlapped(), nullptr);
   wsa_error = status == 0 ? 0 : WSAGetLastError();
   if (wsa_error != 0 && wsa_error != WSA_IO_PENDING) {
     // Async read returned immediately with an error
-    executor_->Run([this, on_read = std::move(on_read), wsa_error]() mutable {
-      on_read(GRPC_WSA_ERROR(
-          wsa_error,
-          absl::StrFormat("WindowsEndpont::%p Read failed", this).c_str()));
+    return GRPC_WSA_ERROR(
+        wsa_error,
+        absl::StrFormat("WindowsEndpont::%p Read failed", this).c_str());
+  }
+  io_state_->socket->NotifyOnRead(&io_state_->handle_read_event);
+  return absl::OkStatus();
+}
+
+bool WindowsEndpoint::Read(absl::AnyInvocable<void(absl::Status)> on_read,
+                           SliceBuffer* buffer, const ReadArgs* /* args */) {
+  GRPC_EVENT_ENGINE_ENDPOINT_TRACE("WindowsEndpoint::%p reading", this);
+  if (io_state_->socket->IsShutdown()) {
+    executor_->Run([on_read = std::move(on_read)]() mutable {
+      on_read(absl::UnavailableError("Socket is shutting down."));
     });
     return false;
   }
+  buffer->Clear();
```

Comment:
That's right. Ideally the buffer should be empty. It would be great if any pre-existing slices in the buffer could be reused, but there's no such contract, and it may be the case that they are in use elsewhere and simply not cleaned out of the SliceBuffer.

That said, I would love to change the `EventEngine::Read` contract to say that any slices in the buffer are now donated to the cause, it could save some allocations. But we'd want to investigate all call sites to see if there is any value in doing so, and we may introduce bugs doing this if slices are left in unintentionally - I don't want to attempt it _yet_.